File: results/2026-04-08_192624/tasks/18-secret-rotation-validator/default-sonnet/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""
Act Test Harness for Secret Rotation Validator
===============================================

Runs the GitHub Actions workflow via `act` for each test case and asserts on:
  1. act exit code == 0  (all jobs succeeded)
  2. Exact expected strings appear in the act output
  3. "Job succeeded" appears in the output

Also runs workflow-structure checks without invoking act:
  - YAML parses correctly and has required triggers / jobs / steps
  - Referenced script / fixture files exist on disk
  - actionlint exits with code 0

All act output is appended to act-result.txt in the current working directory.

Usage:
    python3 run_act_tests.py
"""

import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

# ---------------------------------------------------------------------------
# Paths and configuration
# ---------------------------------------------------------------------------

PROJECT_ROOT = Path(__file__).parent.resolve()
RESULT_FILE = PROJECT_ROOT / "act-result.txt"
WORKFLOW_PATH = PROJECT_ROOT / ".github" / "workflows" / "secret-rotation-validator.yml"

# Platform image — already pulled, so act won't need to download it
ACT_PLATFORM = "ubuntu-latest=catthehacker/ubuntu:act-22.04"

# Files that must be present in every temp repo
PROJECT_FILES = [
    "secret_rotation_validator.py",
    "conftest.py",
    "requirements.txt",
    "tests/__init__.py",
    "tests/test_validator.py",
    ".github/workflows/secret-rotation-validator.yml",
]

# ---------------------------------------------------------------------------
# Fixture data (reference date = 2024-01-15 for all cases)
# ---------------------------------------------------------------------------

FIXTURE_ALL_OK = [
    # 1 day since rotation,  89 days until expiry  →  ok
    {"name": "fresh-key",   "last_rotated": "2024-01-14", "rotation_days": 90, "required_by": ["service-a"]},
    # 5 days since rotation, 85 days until expiry  →  ok
    {"name": "fresh-token", "last_rotated": "2024-01-10", "rotation_days": 90, "required_by": ["service-b"]},
]

FIXTURE_EXPIRED = [
    # 136 days since rotation, -46 days until expiry  →  expired
    {"name": "old-db-password", "last_rotated": "2023-09-01", "rotation_days": 90, "required_by": ["backend-api"]},
]

FIXTURE_WARNING = [
    # 26 days since rotation, 4 days until expiry  →  warning
    {"name": "expiring-api-key", "last_rotated": "2023-12-20", "rotation_days": 30, "required_by": ["frontend"]},
]

FIXTURE_MIXED = [
    {"name": "db-password",  "last_rotated": "2023-09-01", "rotation_days": 90, "required_by": ["backend-api", "worker"]},
    {"name": "api-key",      "last_rotated": "2023-12-20", "rotation_days": 30, "required_by": ["frontend"]},
    {"name": "jwt-secret",   "last_rotated": "2024-01-01", "rotation_days": 90, "required_by": ["auth-service"]},
]

# ---------------------------------------------------------------------------
# Test cases — each specifies fixture data and exact strings to assert
# ---------------------------------------------------------------------------

TEST_CASES = [
    {
        "name": "test_all_ok_markdown",
        "description": "All secrets are fresh — no expired or warnings",
        "fixture": FIXTURE_ALL_OK,
        "expected_strings": [
            "**Expired:** 0",
            "**Warning:** 0",
            "**OK:** 2",
            "fresh-key",
            "fresh-token",
        ],
    },
    {
        "name": "test_expired_markdown",
        "description": "One expired secret appears in the EXPIRED section",
        "fixture": FIXTURE_EXPIRED,
        "expected_strings": [
            "**Expired:** 1",
            "**Warning:** 0",
            "**OK:** 0",
            "old-db-password",
            "## EXPIRED (1)",
            "OVERDUE",
        ],
    },
    {
        "name": "test_warning_markdown",
        "description": "One near-expiry secret appears in the WARNING section",
        "fixture": FIXTURE_WARNING,
        "expected_strings": [
            "**Expired:** 0",
            "**Warning:** 1",
            "**OK:** 0",
            "expiring-api-key",
            "## WARNING (1)",
        ],
    },
    {
        "name": "test_mixed_json",
        "description": "Mixed secrets with JSON output — verify exact counts",
        "fixture": FIXTURE_MIXED,
        "expected_strings": [
            '"total": 3',
            '"expired": 1',
            '"warning": 1',
            '"ok": 1',
            "db-password",
            "api-key",
            "jwt-secret",
        ],
    },
]


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def strip_ansi(text: str) -> str:
    """Remove ANSI escape codes so string searches work reliably."""
    return re.sub(r"\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])", "", text)


def write_result(text: str) -> None:
    """Append text to act-result.txt."""
    with open(RESULT_FILE, "a") as fh:
        fh.write(text)


def setup_temp_repo(fixture_data: list) -> Path:
    """
    Create a temporary git repository populated with all project files and
    the given fixture written to fixtures/secrets_config.json.
    """
    tmpdir = Path(tempfile.mkdtemp(prefix="act-test-"))

    # Copy project files into the temp repo
    for rel_path in PROJECT_FILES:
        src = PROJECT_ROOT / rel_path
        dst = tmpdir / rel_path
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dst)

    # Write test fixture as the default secrets file for the workflow
    fixtures_dir = tmpdir / "fixtures"
    fixtures_dir.mkdir(exist_ok=True)
    (fixtures_dir / "secrets_config.json").write_text(
        json.dumps(fixture_data, indent=2)
    )

    # Initialise git repo (act requires a git repo to simulate push events)
    def git(*args):
        subprocess.run(["git", *args], cwd=tmpdir, check=True, capture_output=True)

    git("init")
    git("symbolic-ref", "HEAD", "refs/heads/main")   # default branch = main
    git("config", "user.email", "test@example.com")
    git("config", "user.name", "Test Runner")
    git("add", "-A")
    git("commit", "-m", "Initial commit")

    return tmpdir


def run_act(tmpdir: Path) -> tuple[int, str]:
    """Run `act push --rm` in tmpdir and return (exit_code, combined_output)."""
    cmd = [
        "act", "push",
        "--rm",
        "-P", ACT_PLATFORM,
        "--pull=false",    # image is already cached locally
    ]
    result = subprocess.run(
        cmd,
        cwd=tmpdir,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        timeout=600,
    )
    return result.returncode, strip_ansi(result.stdout)


# ---------------------------------------------------------------------------
# Workflow structure tests (no act required)
# ---------------------------------------------------------------------------

def test_workflow_structure() -> list[str]:
    """
    Parse the workflow YAML and verify required triggers, jobs, and steps.
    Returns a list of failure messages (empty = all checks passed).
    """
    failures = []

    try:
        import yaml  # PyYAML
    except ImportError:
        # Fall back to a regex-based check if PyYAML is not installed
        text = WORKFLOW_PATH.read_text()
        checks = [
            ("push trigger",             "push:" in text or "push:\n" in text),
            ("pull_request trigger",     "pull_request" in text),
            ("schedule trigger",         "schedule:" in text),
            ("workflow_dispatch trigger","workflow_dispatch" in text),
            ("test job",                 "test:" in text),
            ("validate-markdown job",    "validate-markdown:" in text),
            ("validate-json job",        "validate-json:" in text),
            ("actions/checkout@v4",      "actions/checkout@v4" in text),
            ("actions/setup-python",     "actions/setup-python" in text),
            ("script reference",         "secret_rotation_validator.py" in text),
            ("fixture reference",        "fixtures/secrets_config.json" in text),
        ]
        for name, ok in checks:
            if not ok:
                failures.append(f"Workflow missing: {name}")
        return failures

    wf = yaml.safe_load(WORKFLOW_PATH.read_text())

    # PyYAML parses the bare YAML key `on` as boolean True — handle both forms
    triggers = wf.get("on", wf.get(True, {}))
    for trig in ("push", "pull_request", "schedule", "workflow_dispatch"):
        if trig not in triggers:
            failures.append(f"Missing trigger: {trig}")

    # Jobs
    jobs = wf.get("jobs", {})
    for job in ("test", "validate-markdown", "validate-json"):
        if job not in jobs:
            failures.append(f"Missing job: {job}")

    # validate-* jobs depend on test
    for job in ("validate-markdown", "validate-json"):
        needs = jobs.get(job, {}).get("needs", [])
        if isinstance(needs, str):
            needs = [needs]
        if "test" not in needs:
            failures.append(f"Job '{job}' should depend on 'test'")

    # Script file referenced in workflow steps exists on disk
    for job_def in jobs.values():
        for step in job_def.get("steps", []):
            run_script = step.get("run", "")
            if "secret_rotation_validator.py" in run_script:
                if not (PROJECT_ROOT / "secret_rotation_validator.py").exists():
                    failures.append("secret_rotation_validator.py not found on disk")
                break

    return failures


def test_referenced_files_exist() -> list[str]:
    """Verify all files the workflow references actually exist."""
    failures = []
    required = [
        PROJECT_ROOT / "secret_rotation_validator.py",
        PROJECT_ROOT / "fixtures" / "secrets_config.json",
        PROJECT_ROOT / "requirements.txt",
        PROJECT_ROOT / "tests" / "test_validator.py",
    ]
    for path in required:
        if not path.exists():
            failures.append(f"Required file missing: {path.relative_to(PROJECT_ROOT)}")
    return failures


def test_actionlint() -> tuple[int, str]:
    """Run actionlint on the workflow file and return (exit_code, output)."""
    result = subprocess.run(
        ["actionlint", str(WORKFLOW_PATH)],
        capture_output=True,
        text=True,
    )
    return result.returncode, result.stdout + result.stderr


# ---------------------------------------------------------------------------
# Main test runner
# ---------------------------------------------------------------------------

def main() -> int:
    # Start fresh
    RESULT_FILE.unlink(missing_ok=True)

    total_failures = 0

    # -----------------------------------------------------------------------
    # Section A: Workflow structure tests
    # -----------------------------------------------------------------------
    write_result("=" * 70 + "\n")
    write_result("SECTION A: Workflow Structure Tests\n")
    write_result("=" * 70 + "\n\n")

    print("=== Section A: Workflow Structure Tests ===")

    # A1: structure check
    struct_failures = test_workflow_structure()
    if struct_failures:
        for f in struct_failures:
            msg = f"  FAIL: {f}\n"
            print(msg, end="")
            write_result(msg)
        total_failures += len(struct_failures)
    else:
        msg = "  PASS: workflow structure (triggers, jobs, dependencies)\n"
        print(msg, end="")
        write_result(msg)

    # A2: referenced files exist
    file_failures = test_referenced_files_exist()
    if file_failures:
        for f in file_failures:
            msg = f"  FAIL: {f}\n"
            print(msg, end="")
            write_result(msg)
        total_failures += len(file_failures)
    else:
        msg = "  PASS: all referenced files exist on disk\n"
        print(msg, end="")
        write_result(msg)

    # A3: actionlint
    lint_code, lint_output = test_actionlint()
    write_result(f"\n--- actionlint output ---\n{lint_output or '(no output)'}\n")
    if lint_code != 0:
        msg = f"  FAIL: actionlint exited {lint_code}\n"
        print(msg, end="")
        write_result(msg)
        total_failures += 1
    else:
        msg = "  PASS: actionlint (exit code 0)\n"
        print(msg, end="")
        write_result(msg)

    write_result("\n")

    # -----------------------------------------------------------------------
    # Section B: Act execution tests
    # -----------------------------------------------------------------------
    write_result("=" * 70 + "\n")
    write_result("SECTION B: Act Execution Tests\n")
    write_result("=" * 70 + "\n\n")

    print("\n=== Section B: Act Execution Tests ===")

    for tc in TEST_CASES:
        name = tc["name"]
        print(f"\n--- {name}: {tc['description']} ---")
        write_result(f"{'=' * 70}\n")
        write_result(f"TEST CASE: {name}\n")
        write_result(f"Description: {tc['description']}\n")
        write_result(f"Fixture: {json.dumps(tc['fixture'], indent=2)}\n\n")

        # Set up isolated temp repo
        tmpdir = setup_temp_repo(tc["fixture"])
        case_failures = 0

        try:
            print(f"  Running act push --rm in {tmpdir} ...")
            exit_code, output = run_act(tmpdir)

            write_result(f"--- act output ---\n{output}\n\n")

            # Assertion 1: exit code must be 0
            if exit_code == 0:
                write_result("ASSERT exit_code == 0: PASS\n")
                print("  PASS: act exited with code 0")
            else:
                write_result(f"ASSERT exit_code == 0: FAIL (got {exit_code})\n")
                print(f"  FAIL: act exited with code {exit_code}")
                case_failures += 1

            # Assertion 2: "Job succeeded" appears for each job
            job_succeeded_count = output.lower().count("job succeeded")
            if job_succeeded_count >= 3:
                write_result(f"ASSERT 'Job succeeded' x3: PASS ({job_succeeded_count} occurrences)\n")
                print(f"  PASS: found 'Job succeeded' {job_succeeded_count} times")
            else:
                write_result(f"ASSERT 'Job succeeded' x3: FAIL (found {job_succeeded_count})\n")
                print(f"  FAIL: expected 3 'Job succeeded', got {job_succeeded_count}")
                case_failures += 1

            # Assertion 3: exact expected strings
            for expected in tc["expected_strings"]:
                if expected in output:
                    write_result(f"ASSERT '{expected}': PASS\n")
                    print(f"  PASS: found expected string: {expected!r}")
                else:
                    write_result(f"ASSERT '{expected}': FAIL (not found in output)\n")
                    print(f"  FAIL: expected string not found: {expected!r}")
                    case_failures += 1

        except subprocess.TimeoutExpired:
            write_result("ERROR: act timed out after 600s\n")
            print("  ERROR: act timed out")
            case_failures += 1
        except Exception as exc:
            write_result(f"ERROR: {exc}\n")
            print(f"  ERROR: {exc}")
            case_failures += 1
        finally:
            shutil.rmtree(tmpdir, ignore_errors=True)

        if case_failures == 0:
            write_result(f"\nRESULT: PASS (all assertions passed)\n\n")
            print(f"  RESULT: PASS")
        else:
            write_result(f"\nRESULT: FAIL ({case_failures} assertion(s) failed)\n\n")
            print(f"  RESULT: FAIL ({case_failures} failures)")
            total_failures += case_failures

    # -----------------------------------------------------------------------
    # Final summary
    # -----------------------------------------------------------------------
    summary = (
        f"\n{'=' * 70}\n"
        f"FINAL RESULT: {'PASS' if total_failures == 0 else 'FAIL'} "
        f"({total_failures} total failure(s))\n"
        f"{'=' * 70}\n"
    )
    write_result(summary)
    print(summary)

    return 0 if total_failures == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/typescript-bun-sonnet46-1m-medium/generated-code/run-act-tests.ts
// Test harness: runs all test cases through the GitHub Actions workflow via `act push`.
// Also includes workflow structure tests (YAML checks, actionlint, file existence).
//
// Usage: bun run run-act-tests.ts
// Output: act-result.txt in the current working directory

import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { execSync, spawnSync } from "child_process";

const PROJECT_ROOT = path.dirname(new URL(import.meta.url).pathname);
const ACT_RESULT_FILE = path.join(PROJECT_ROOT, "act-result.txt");
const WORKFLOW_FILE = path.join(PROJECT_ROOT, ".github/workflows/semantic-version-bumper.yml");

// Clear act-result.txt at the start
fs.writeFileSync(ACT_RESULT_FILE, "");

let passCount = 0;
let failCount = 0;

function pass(name: string): void {
  console.log(`  PASS: ${name}`);
  passCount++;
}

function fail(name: string, reason: string): never {
  console.error(`  FAIL: ${name} — ${reason}`);
  failCount++;
  throw new Error(`Assertion failed: ${name} — ${reason}`);
}

function assert(condition: boolean, name: string, reason: string): void {
  if (!condition) fail(name, reason);
  else pass(name);
}

function assertContains(haystack: string, needle: string, name: string): void {
  assert(haystack.includes(needle), name, `Expected to find: "${needle}"`);
}

function appendActResult(header: string, output: string): void {
  const divider = "=".repeat(70);
  fs.appendFileSync(
    ACT_RESULT_FILE,
    `\n${divider}\n${header}\n${divider}\n${output}\n`
  );
}

// ---------------------------------------------------------------------------
// SECTION 1: Workflow structure tests (no act needed, fast)
// ---------------------------------------------------------------------------

console.log("\n=== Workflow Structure Tests ===\n");

// 1a. Workflow file exists
assert(fs.existsSync(WORKFLOW_FILE), "workflow file exists", `${WORKFLOW_FILE} not found`);

const workflowContent = fs.readFileSync(WORKFLOW_FILE, "utf-8");

// 1b. Has push trigger
assertContains(workflowContent, "push:", "workflow has push trigger");

// 1c. Has pull_request trigger
assertContains(workflowContent, "pull_request:", "workflow has pull_request trigger");

// 1d. Has workflow_dispatch trigger
assertContains(workflowContent, "workflow_dispatch:", "workflow has workflow_dispatch trigger");

// 1e. Has test-and-bump job
assertContains(workflowContent, "test-and-bump:", "workflow has test-and-bump job");

// 1f. Uses actions/checkout@v4
assertContains(workflowContent, "actions/checkout@v4", "workflow uses actions/checkout@v4");

// 1g. Runs bun test
assertContains(workflowContent, "bun test", "workflow runs bun test");

// 1h. References correct script path
assertContains(workflowContent, "src/version-bumper.ts", "workflow references version-bumper.ts");

// 1i. References fixture files
assertContains(workflowContent, "fixtures/test-case-patch.json", "workflow references patch fixture");
assertContains(workflowContent, "fixtures/test-case-minor.json", "workflow references minor fixture");
assertContains(workflowContent, "fixtures/test-case-major.json", "workflow references major fixture");

// 1j. Script files actually exist
assert(
  fs.existsSync(path.join(PROJECT_ROOT, "src/version-bumper.ts")),
  "src/version-bumper.ts exists",
  "Script file not found"
);
assert(
  fs.existsSync(path.join(PROJECT_ROOT, "fixtures/test-case-patch.json")),
  "fixtures/test-case-patch.json exists",
  "Patch fixture not found"
);
assert(
  fs.existsSync(path.join(PROJECT_ROOT, "fixtures/test-case-minor.json")),
  "fixtures/test-case-minor.json exists",
  "Minor fixture not found"
);
assert(
  fs.existsSync(path.join(PROJECT_ROOT, "fixtures/test-case-major.json")),
  "fixtures/test-case-major.json exists",
  "Major fixture not found"
);

// 1k. actionlint passes
console.log("\n  Running actionlint...");
const actionlintResult = spawnSync("actionlint", [WORKFLOW_FILE], {
  encoding: "utf-8",
  cwd: PROJECT_ROOT,
});
assert(actionlintResult.status === 0, "actionlint passes", actionlintResult.stdout + actionlintResult.stderr);

appendActResult("WORKFLOW STRUCTURE TESTS", "All workflow structure tests passed.");

// ---------------------------------------------------------------------------
// SECTION 2: act end-to-end test (one act push run covers all 3 fixtures)
// ---------------------------------------------------------------------------

console.log("\n=== Act End-to-End Tests ===\n");

// Build test data: what each step should output
interface ActTestCase {
  stepName: string;
  expectedVersion: string;
}

const actTestCases: ActTestCase[] = [
  {
    stepName: "Run version bumper (patch fixture)",
    expectedVersion: "1.0.1",
  },
  {
    stepName: "Run version bumper (minor fixture)",
    expectedVersion: "1.1.0",
  },
  {
    stepName: "Run version bumper (major fixture)",
    expectedVersion: "2.0.0",
  },
];

// Set up a temp git repo with all project files
function setupTempRepo(): string {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "ver-bump-act-"));

  // Directories to copy
  const dirs = ["src", "fixtures", ".github"];
  for (const dir of dirs) {
    copyDirSync(path.join(PROJECT_ROOT, dir), path.join(tmpDir, dir));
  }

  // Files to copy
  const files = ["package.json", ".actrc"];
  for (const file of files) {
    const src = path.join(PROJECT_ROOT, file);
    if (fs.existsSync(src)) {
      fs.copyFileSync(src, path.join(tmpDir, file));
    }
  }

  // Init git repo
  execSync(
    "git init && " +
      "git config user.email 'test@test.com' && " +
      "git config user.name 'Test' && " +
      "git add -A && " +
      "git commit -m 'test: initial commit'",
    { cwd: tmpDir, stdio: "pipe" }
  );

  return tmpDir;
}

function copyDirSync(src: string, dest: string): void {
  fs.mkdirSync(dest, { recursive: true });
  for (const entry of fs.readdirSync(src, { withFileTypes: true })) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);
    if (entry.isDirectory()) {
      copyDirSync(srcPath, destPath);
    } else {
      fs.copyFileSync(srcPath, destPath);
    }
  }
}

console.log("  Setting up temp git repo...");
const tmpRepo = setupTempRepo();
console.log(`  Temp repo: ${tmpRepo}`);

console.log("  Running act push --rm (this takes 30-90 seconds)...");

const actResult = spawnSync("act", ["push", "--rm", "--pull=false"], {
  cwd: tmpRepo,
  encoding: "utf-8",
  timeout: 300_000, // 5 minutes
});

const actOutput = (actResult.stdout ?? "") + (actResult.stderr ?? "");

appendActResult(
  "ACT PUSH RUN — all three fixture test cases",
  actOutput
);

// Assert act exit code 0
assert(
  actResult.status === 0,
  "act push exits with code 0",
  `act exited with ${actResult.status}. Output:\n${actOutput.slice(-2000)}`
);

// Assert "Job succeeded"
assertContains(actOutput, "Job succeeded", "act reports Job succeeded");

// Assert exact expected versions for each test case
for (const tc of actTestCases) {
  assertContains(
    actOutput,
    `New version: ${tc.expectedVersion}`,
    `${tc.stepName} outputs "New version: ${tc.expectedVersion}"`
  );
}

// Clean up temp dir
fs.rmSync(tmpRepo, { recursive: true, force: true });

// ---------------------------------------------------------------------------
// Summary
// ---------------------------------------------------------------------------

console.log(`\n=== Results: ${passCount} passed, ${failCount} failed ===\n`);
console.log(`act-result.txt written to: ${ACT_RESULT_FILE}`);

if (failCount > 0) {
  process.exit(1);
}
